core/src/main/scala/kafka/log/Log.scala

```diff
@@ -717,16 +746,20 @@ class Log(@volatile var dir: File,
     deleteOldSegments(shouldDelete)
   }
 
+  /*
+   * The log's next offset may be smaller than log start offset when the log is purged upon PurgeRequest
+   * We don't need to call this method periodically using scheduler since it is called right after broker receives PurgeRequest
+   */
+  private def deleteLogStartOffsetBreachedSegments() : Int = {
+    // keep active segment to avoid frequent log rolling due to user's PurgeRequest
+    deleteOldSegments(segment => segment.nextOffset <= logStartOffset && segment.baseOffset < activeSegment.baseOffset)
```

Comment:
Discussed offline. The issue is resolved after we change the definition of lowWatermark to be the minimum logStartOffset of all live replicas.